File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/chat/PublicChatActor.java
package net.bestia.zoneserver.actor.chat;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.entity.Entity;
import net.bestia.entity.EntityService;
import net.bestia.entity.component.PositionComponent;
import net.bestia.messages.chat.ChatMessage;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.actor.zone.SendClientsInRangeActor;
import net.bestia.zoneserver.entity.PlayerEntityService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;
import java.util.Optional;

/**
 * Handles public chat of the user and sends them to all entities which can
 * receive them.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class PublicChatActor extends AbstractActor {

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);
	public static final String NAME = "public";

	private final PlayerEntityService playerEntityService;
	private final EntityService entityService;
	private final ActorRef sendActiveRange;

	@Autowired
	public PublicChatActor(PlayerEntityService playerEntityService,
			EntityService entityService) {
		
		this.playerEntityService = Objects.requireNonNull(playerEntityService);
		this.entityService = Objects.requireNonNull(entityService);
		this.sendActiveRange = SpringExtension.actorOf(getContext(), SendClientsInRangeActor.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(ChatMessage.class, this::handlePublic)
				.build();
	}

	/**
	 * Sends a public message to all clients in sight.
	 */
	private void handlePublic(ChatMessage chatMsg) {
		// Sanity check.
		if(chatMsg.getChatMode() != ChatMessage.Mode.PUBLIC) {
			LOG.warning("Can not handle non public chat messages: {}.", chatMsg);
			unhandled(chatMsg);
			return;
		}
		
		final long accId = chatMsg.getAccountId();
		final Entity pbe = playerEntityService.getActivePlayerEntity(accId);

		if (pbe == null) {
			return;
		}

		// Add the current entity id to the message.
		final ChatMessage chatEntityMsg = new ChatMessage(accId, pbe.getId(), chatMsg);

		Optional<PositionComponent> pos = entityService.getComponent(pbe, PositionComponent.class);

		if (!pos.isPresent()) {
			LOG.warning("Player bestia has no position component.");
			return;
		}

		// We dont need to send a echo back because the player entity is also
		// active in the area so this call also includes the sender of the chat
		// message.
		sendActiveRange.tell(chatEntityMsg, getSelf());
	}

}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/chat/PublicChatActor.kt
package net.bestia.zoneserver.actor.chat

import akka.actor.AbstractActor
import net.bestia.messages.chat.ChatRequest
import net.bestia.zoneserver.actor.routing.MessageApi
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.actor.client.SendClientsInRangeActor
import net.bestia.zoneserver.actor.client.SendInRange
import net.bestia.zoneserver.actor.entity.awaitEntityResponse
import net.bestia.zoneserver.chat.PublicChatService
import java.lang.IllegalStateException

/**
 * Handles public chat of the user and sends them to all entities which can
 * receive them.
 *
 * @author Thomas Felix
 */
@Actor
class PublicChatActor(
    private val publicChatService: PublicChatService,
    private val messageApi: MessageApi
) : AbstractActor() {

  private val sendActiveRange = SpringExtension.actorOf(context, SendClientsInRangeActor::class.java)

  override fun createReceive(): AbstractActor.Receive {
    return receiveBuilder()
        .match(ChatRequest::class.java, this::handlePublic)
        .build()
  }

  /**
   * Sends a public message to all clients in sight.
   */
  private fun handlePublic(chat: ChatRequest) {
    val response = publicChatService.getChatResponse(chat)
    val entityId = response.entityId ?: throw IllegalStateException("Player has not active Bestia")

    awaitEntityResponse(messageApi, context, entityId) {
      val sendInRange = SendInRange(it, response)

      // We dont need to send a echo back because the player entity is also
      // active in the area so this call also includes the sender of the chat
      // message.
      sendActiveRange.tell(sendInRange, self)
    }
  }

  companion object {
    const val NAME = "public"
  }
}
